File: Python/textbook_track/chapter08/ba8a.py
# Implement FarthestFirstTraversal
import functools as ft
import math
import sys


def convert_to_numbers(line, converter_func):
    return [converter_func(elem) for elem in line.split()]


def read_points(reader):
    return [convert_to_numbers(line, float) for line in reader]


def euclidean_distance(p, q):
    return math.sqrt(sum((p_i - q_i)**2 for p_i, q_i in zip(p, q)))


def distance_from_centers(point, centers):
    distance_from_point = ft.partial(euclidean_distance, q=point)
    return min(map(distance_from_point, centers))


def find_index_that_maximizes_distance(points, center_indices):
    centers = [points[ix] for ix in center_indices]
    max_ix = 0
    max_distance = 0.0
    for ix, point in enumerate(points):
        if ix not in center_indices:
            d = distance_from_centers(point, centers)
            if d >= max_distance:
                max_distance = d
                max_ix = ix
    return max_ix


def farthest_first_traversal(points, k):
    center_indices = [0]
    for _ in range(k-1):
        ix = find_index_that_maximizes_distance(points, center_indices)
        center_indices.append(ix)
    return [points[ix] for ix in center_indices]


def main():
    reader = sys.stdin
    k, _ = convert_to_numbers(next(reader), int)
    points = read_points(reader)
    result = farthest_first_traversal(points, k)
    for point in result:
        print(' '.join(map(str, point)))


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter08/BA8A.scala
package TextbookTrack.Chapter08

object BA8A {
  import scala.reflect.ClassTag

  type Point = List[Double]

  private def convertToNumber[T: ClassTag](line: String, converter: String => T): List[T] =
    line.split(" ").map(converter).toList

  private def readParameters(line: String): (Int, Int) = convertToNumber(line, _.toInt) match {
    case List(k, m) => (k, m)
    case _ => throw new Exception("Unexpected number of parameters.")
  }

  private def euclideanDistance(p: Point, q: Point): Double =
    math.sqrt(p.lazyZip(q).foldLeft(0.0){ case (d, (pi, qi)) => d + math.pow(pi - qi, 2) })

  private def distanceFromCenters(point: Point, centers: List[Point]): Double =
    centers.map(euclideanDistance(_, point)).min

  private def findIndexThatMaximizesDistance(points: List[Point], centerIndices: List[Int]): Int = {
    val centers: List[Point] = centerIndices.map(points)
    val (maxIx, _): (Int, Double) = points.zipWithIndex.foldLeft((0, 0.0)){
      case (acc @ (_, maxDistance), (point, ix)) =>
        if (centerIndices.contains(ix)) acc
        else {
          val d: Double = distanceFromCenters(point, centers)
          if (d >= maxDistance) (ix, d) else acc
        }
    }
    maxIx
  }

  def farthestFirstTraversal(points: List[Point], k: Int): List[Point] = {
    val centerIndices: List[Int] = (0 until k - 1).foldLeft(List(0)){
      case (indices, _) => findIndexThatMaximizesDistance(points, indices) :: indices
    }
    centerIndices.map(points).reverse
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val (k, _): (Int, Int) = readParameters(reader.next())
    val points: List[Point] = reader.map(convertToNumber(_, _.toDouble)).toList
    val result: List[Point] = farthestFirstTraversal(points, k)
    result.foreach(point => println(point.mkString(" ")))
  }
}
